compiler/src/dotty/tools/dotc/ast/Trees.scala

```diff
@@ -19,6 +19,76 @@ import language.implicitConversions
 
 object Trees {
 
+  type TreeTag = Int
+
+  object Tag {
+    final val Ident = 0
+    final val Select = 1
+    final val This = 2
+    final val Super = 3
+    final val Apply = 4
+    final val TypeApply = 5
+    final val Literal = 6
+    final val New = 7
+    final val Typed = 8
+    final val NamedArg = 9
+    final val Assign = 10
+    final val Block = 11
+    final val If = 12
+    final val Closure = 13
+    final val Match = 14
+    final val CaseDef = 15
+    final val Return = 16
+    final val Try = 17
+    final val SeqLiteral = 18
+    final val Inlined = 19
+    final val TypeTree = 20
+    final val SingletonTypeTree = 21
+    final val AndTypeTree = 22
+    final val OrTypeTree = 23
+    final val RefinedTypeTree = 24
+    final val AppliedTypeTree = 25
+    final val LambdaTypeTree = 26
+    final val ByNameTypeTree = 27
+    final val TypeBoundsTree = 28
+    final val Bind = 29
+    final val Alternative = 30
+    final val UnApply = 31
+    final val ValDef = 32
+    final val DefDef = 33
+    final val TypeDef = 34
+    final val Template = 35
+    final val Import = 36
+    final val PackageDef = 37
+    final val Annotated = 38
+    final val Thicket = 39
+
+    final val NumTypedTreeTags = 40
+
+    final val TypedSplice = 40
+    final val ModuleDef = 41
+    final val ParsedTry = 42
```

Comment:
I would add a comment describing which tag numbers can be reused and why.

Reply:
👍 It's not clear for me why are they reused.

Reply:
There's no re-use: NumTypeTreeTags is a counter, not a tag. I'll add a comment to make that clearer.